exercism/python/pangram/pangram.py

```diff
@@ -1,3 +1,111 @@
+# Initial Solution - Use a string
+from string import ascii_lowercase as lowercase_letters
+
+
+def is_pangram(sentence):
+    letters = ""
+
+    for char in sentence.lower():
+        if char.isalpha() and char not in letters:
+            letters += char
+    return len(letters) == len(lowercase_letters)
+
+"""
+Pros: Simple to read and understand
+Cons: Not efficient in terms of complexity. That's because strings are immutable in Python. Each time I concatentate strings, an entirely new string is created and the letter is added to it.
+"""
+
+
+###Solutions When Apply Looking Back Approaches###
+
+# Approach 1: Solve problem with a different approach
+# Approach 1 Solution - ASCII representation
+from string import ascii_lowercase as lowercase_letters
+
+ALPHA_LENGTH = 26
+
+
+def is_pangram(sentence):
+    total_letters = []
+    for i in range(ALPHA_LENGTH):
+        total_letters.append(0)
+
+    for char in sentence.lower():
+        if char.isalpha():
+            letter_index = ord(char) - ord("a")
+            total_letters[letter_index] = 1
+
+    return sum(total_letters) == ALPHA_LENGTH
+
+"""
+Pros: Relatively simple to understand; more efficient in terms of time/space complexity; creative solution 
+Cons: It's a bit verbose 
+"""
+
+
+
+# Approach 2: Add a contraint
+# Approach 2 Solution - Solve the problem without using a data structure in the function
+from string import ascii_lowercase as lowercase_letters
+
+
+def is_pangram(sentence):
+    for letter in lowercase_letters:
+        if letter not in sentence.lower():
```

Comment:
You are calling `sentence.lower()` 26 times...you can extract this to a variable, such as `lower_sentence` and call it once before the loop.

Reply:
Great point!

I made that change. I also see that I can extract sentence.lower() into a variable elsewhere, including this solution:

Make sense to do here here, too, right?

```
def is_pangram(sentence):
    lowercase_sentence = sentence.lower()
    
    for letter in lowercase_letters:
        if letter not in lowercase_sentence:
            return False
    return True
```

Reply:
I'm sorry, I think I was mistaken on this. It's only called once.